docs/proposals-accepted/202206-block-series-deletion-compactor.md

```diff
@@ -0,0 +1,206 @@
+* **start timestamp**
+* **end timestamp**: end timestamp should be the tombstone creation time if not specified, otherwise the compactor needs to wait for a long time to make sure all matched blocks are processed. Considering the tombstone pending period, the end timestamp can be set to `(now + pending period)` as well. We don't consider history blocks backfilling now.
+* **author**
+* **reason**
+
+### Store Gateway Masks Series based on Tombstones
+
+Like block metadata, the store gateway also fetches tombstones in a configurable interval, which means some delay is expected: user creates a tombstone but she is still able to query the data to be deleted.
+
+There are different ways to mask series based on tombstones. We will discuss two options here:
+
+#### Mask Series At Query Time
+
+During query time, the store gateway decides what blocks to query based on the request min, max time and label matchers. Meanwhile, it can also decide what tombstones are going to apply for the selected blocks based on tombstone time range and label matchers.
+
+Then during the query time, it checks the series label sets against the tombstone matchers, as well as the chunks and tombstone time range to mask the series data.
+
+Pros:
+* Easy to implement and very straightforward
+* Less memory usage.
+
+Cons:
+* Slower at query time
+
+#### Mask Series At Tombstone Sync Time
+
+During the tombstone file sync, store gateway can also eagerly apply the tombstones to all blocks, calculate the matched postings and time range for each block. For each block, each tombstone's processed result will be stored in memory using the Prometheus `MemTombstone` [format](https://github.com/prometheus/prometheus/blob/main/tsdb/tombstones/tombstones.go#L230).
+
+```go
+// Result for the processed result for each tombstone, each block.
+type MemTombstones struct {
+    intvlGroups map[storage.SeriesRef]Intervals
+    mtx         sync.RWMutex
+}
+
+// Result for the processed result for each block.
+// Key: global tombstone ID, Value: processed result for each tombstone.
+type BlockTombstones map[ulid.ULID]*MemTombstones
+```
+
+Since all the blocks and tombstones are immutable, this result can be easily cached in memory or in filesystem. Ideally for each tombstone, each block we just need to calculate once. Since `BlockTombstones` is a map, it is easy to add or delete one key if users want to add more tombstones or remove one tombstone.
+
+During the query time, the cached tombstone will be used. As it knows which `SeriesRef`, which intervals to apply, it is very straightforward to mask series.
+
+Pros:
+* More efficient during query time, less processing required.
+* As data can be cached there is not much computation required.
+
+Cons:
+* More memory required.
+* If read is much less than write, it is inefficient as the pre-calculated result won't be used much.
+
+From my perspective, option 2 is better because data can be cached.
```

Comment:
As I reader, I wonder what is `option 2` here. You mentioned "options" before, but you didn't number them 1 or 2. I am confused whether this is a reference to `Option 2: New API to Match Tombstones` (inside the tombstone deletion heading). 

I recommend numbering the options `Mask Series At Query Time` and `Mask Series At Tombstone Sync Time`, like done for the discussions about deletion with Compactor.

I believe that putting your recommendation under a separate heading would be great to give it evidence.

Reply:
Thanks! That makes total sense and I should really do it from the beginning. 💯